docs/designs/ps_design.md

```diff
@@ -1,375 +1,191 @@
-        pass
+### KVStore
 
-    def get_embedding_vector(self, name, indices):
-        pass
+The KVStore needs to support both embedding vector parameter and non-embedding parameter.
 
-    def set_parameter(self, name, value):
-        pass
+Since `tf.keras.optimizer` only accept `tf.Variable` as the parameter. To avoid unnecessary memory copy, we save non-embedding parameter as a `tf.Variable` in variable DB directly.
 
-    def set_embedding_vector(self, name, indices, value):
-        pass
-```
+For the embedding vector parameter, there are two reasons that we could not initialize it before training:
+
+- Feature ID is usually calculated by the hash function. The training dataset is too large to get the range of the ID before training.
+- Sometimes, there comes a new feature ID in online learning.
 
-For a common model variable, we use save it as a `tf.Variable` in Parameter DB. For the embedding table, we introduce a customized data structure.
+We introduce a customized data structure `EmbeddingVectors`, Following is the definition of `EmbeddingVectors`:
 
 ```python
-class EmbeddingTable(object):
+class EmbeddingVectors(object):
     def __init__(self, name, dim, initializer):
         self.name = name
         self.dim = dim
         self.initializer = initializer
         self.vectors = {}
 
     def get(self, indices):
-        pass
+        res = []
+        for i in indices:
+            if i not in self.vectors:
+                value = init_value(self.initializer, self.dim)
+            else:
+                value = self.vectors[i]
+            res.append(value)
+        return res
 
     def set(self, indices, value):
         pass
 ```
 
-The name of embedding table is the embedding layer name. EmbeddingTable uses a dictionary `vectors` to store `<id, embedding_vector>` pairs.
+The name is the embedding layer name of embedding vectors. It uses a dictionary `vectors` to store embedding vector.
 
-Since embedding table is lazily initialized in PS, it also has `dim` and `initializer` fields.
+Since embedding vectors are lazily initialized in PS, it also has `dim` and `initializer` fields. Inside the `get` interface of `EmbeddingVectors `, if the id is not in the `vectors` dictionary, the corresponding value will be initialized and return back.
 
-### Tensor Data Structure
-
-To support data communication between workers and PS, we introduce a `Tensor` proto message:
-
-```proto
-message Tensor {
-    enum DataType {
-        BOOL = 0;
-        INT16 = 1;
-        INT32 = 2;
-        INT64 = 3;
-        FP16 = 4;
-        FP32 = 5;
-        FP64 = 6;
-    }
-    string name = 1;
-    DataType data_type = 2;
-    repeated int64 dim = 3;
-    bytes content = 4;
-    repeated int64 indices = 5;
-}
-```
-
-Correspondingly, we have a `Tensor` Python class.
+There could be multiple embedding vectors from different embedding layer. We save all these embedding vectors in an independent embedding vector DB of KVStore.
```

Comment:
It means the KVStore contains multiple embedding vectors of different embedding layer.
How does KVStore figure out which embedding vector group to look up?

Reply:
There are two steps: 

- give embedding layer name, `KVStore` returns a `EembeddingVectors` object
- give ids info, `EembeddingVectors` returns embedding vectors